### PR TITLE
[ROCM] Patch to enable rocm for r2.0 release branch

### PR DESCRIPTION
**Pull request by sunway513**
The goal for this PR is to patch Tensorflow r2.0 release, so it would fully enable ROCm
non-xla path support.
Most of the PRs been cherry-picked in this patch have already been upstreamed in the
upstream master branch.
cc @deven-amd @whchung

The following were all the related commits been cherry-picked:

Commits on Aug 20, 2019
deven-amd and sunway513
adding/updating ROCm support in the ci_build scripts
d5a0eee
deven-amd and sunway513
updating Dockerfile.rocm to pick a specific version of the rocm libra… …
e335575
deven-amd and sunway513
adding a script for testing the ROCm Community Supported Build
ae83a20

Commits on Aug 22, 2019
deven-amd and sunway513
Resolve merge conflicts for PR #31393
73ff708
deven-amd and sunway513
The following PR/commit breaks the --config=rocm build …
614bdb5
deven-amd and sunway513
updating testcases to work correctly with ROCm
1685240
jeffdaily and sunway513
improve concurrency between compute and nccl streams …
3fbb049
whchung and sunway513
[ROCm] enable roll op on ROCm.
1d5f440
whchung and sunway513
[ROCm] enable InTopK op on ROCm.
941f713
deven-amd and sunway513
updating README.md with information on ROCm Community Supported Builds
73ce64e

Commits on Aug 25, 2019
houtoms and sunway513
fixed potential rocm breaks from use_padded_io
0832b33
deven-amd and sunway513
adding no_rocm tag on unit-tests that check features that are current… …
7aed626
deven-amd and sunway513
Adding ROCm support for reduction ops
82bd216
sunway513
Fix ROCm path build error in rocm_dnn.h
5dba305

Commits on Aug 27, 2019
deven-amd
fixing test failures by skipping parts that functionality not yet sup… …
be6378c
sunway513
Merge pull request #616 from ROCmSoftwarePlatform/r2.0-rocm-upstream-… …
d98a943
sunway513
Add no_rocm tag to //tensorflow/python:stateful_random_ops_test_gpu
d05a47f

Commits on Sep 04, 2019
sunway513
Merge branch 'r2.0-rocm-upstream' of https://github.com/ROCmSoftwareP… …
b1148e4

Commits on Sep 06, 2019
deven-amd and sunway513
adding ROCm support in the build_pip_package script
b908324

### CODE REVIEW

**Comment by sunway513** [comment]
Hi @martinwicke , the basic functionality for ROCm has been enabled in the r2.0 release branch, e.g. we can pass the basic TF 2.0 mnist sample using the docker container built with tensorflow r2.0 branch out of the box.
However, there're 20ish unit tests failures for the test cases defined in the following script:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/linux/rocm/run_csb_tests.sh

The essential goal for this PR is to cherry-pick several missing PRs (already been upstreamed to tensorflow master branch) to the r2.0 release branch, so we can enable additional operators, e.g. reduction ops with commit https://github.com/tensorflow/tensorflow/commit/82bd216ebd98c6c85a221adadd8d5dee902e3f1d, as well as fixing the failed 20ish test cases.

File: tensorflow/core/kernels/reduction_ops.h [hunk 1]
```diff
@@ -117,6 +117,10 @@ struct Identity {
 FIX_MEAN_IDENTITY(Eigen::half)
 FIX_MEAN_IDENTITY(float)
 FIX_MEAN_IDENTITY(double)
+#if GOOGLE_CUDA
+FIX_MEAN_IDENTITY(complex64)
```
Comment (chsigg) [inline tensorflow/core/kernels/reduction_ops.h]: Please revert this change.

File: tensorflow/core/nccl/nccl_manager.cc [hunk 2]
```diff
@@ -533,7 +533,7 @@ void NcclManager::RunCollective(Collective* collective) {
       // Wait to ensure that the kernel that produces the data in the input
       // tensor has finished running before the nccl kernel runs on the
       // communication stream.
-      nccl_stream->stream->ThenWaitFor(p->tensor_stream);
+      nccl_stream->stream->ThenWaitFor(p->input_event.get());
```
Comment (chsigg) [inline tensorflow/core/nccl/nccl_manager.cc]: Please exclude the corresponding commit from this PR. It is too significant for a cherry-pick.

Comment (sunway513) [reply tensorflow/core/nccl/nccl_manager.cc]: Reverted in https://github.com/tensorflow/tensorflow/pull/32474/commits/41041a9c7920a8eeaad0f8feff2838649b49d5ac

File: tensorflow/core/kernels/reduction_ops.h (hunk 1 above)
Comment (sunway513) [reply tensorflow/core/kernels/reduction_ops.h]: Reverted in https://github.com/tensorflow/tensorflow/pull/32474/commits/2118059a0f79fc14e14631663bf3fc2ecbe2bad9

**Comment by sunway513** [comment]
Thanks @chsigg , we have addressed your comments with commit 2118059 and 41041a9, can you review the changes again?

**Review by chsigg — DISMISSED** [review]
This does look safe to merge now. But unfortunately it is awfully late to make it into the 2.0 release.
I will contact @goldiegadde, she owns this release.

**Comment by sunway513** [comment]
Hi @chsigg , gentle ping.

**Comment by martinwicke** [comment]
2.0 itself will not be built with this, it is too late for that. However, we can accept the cherry-pick once we have 2.0 binaries finalized, so that if people build from the 2.0 branch, they get ROCm capabilities.

@goldiegadde does that sound sensible?

**Comment by mihaimaruseac** [comment]
Now that 2.0 is launched, I think we can merge this. CC @goldiegadde @martinwicke

**Review by chsigg — APPROVED** [review]

**Comment by sunway513** [comment]
Now we have all checks passed, can we merge the PR now? 
Gentle ping @martinwicke @goldiegadde

**Comment by goldiegadde** [comment]
> Now we have all checks passed, can we merge the PR now?
> Gentle ping @martinwicke @goldiegadde

Thanks @sunway513 for the reminder. I am merging this PR now, but we are not going to publish a new release with this, the branch of course will have the changes.
